### PR TITLE
Overview and Resource Safety section of documentation

### PR DESCRIPTION
**Pull request by Avasil**
#113 

If you want to build the website locally:

```
sbt docs/docusaurusCreateSite
website/yarn start
```

I will have to run new benchmarks since I mention them and there were some changes in competing libraries

cc @Executioner1939

### CODE REVIEW

File: docs/overview.md [hunk 1]
```diff
@@ -3,4 +3,181 @@ id: overview
...
+To me, the big difference between Monix and other effect libraries is approach to impure code.
+Both `IO` and `ZIO` will push you to write 100% purely functional codebase, except for isolated cases where low-level imperative code is needed for performance.
+Monix favors purely functional programming too (the only impure method is `memoize`), but we also recognize and fully support users who would rather go for hybrid approach, or are allergic to pure FP.
+I feel like it plays to Scala's unique strengths. 
+Few examples of Monix providing extra support for `Future` users:
+- `monix-execution` module provides many utilities to use with `Future` even if you're not interested in `Task` at all.
+- Monix uses `Scheduler` which is also `ExecutionContext` and can be used with `Future` directly. 
+- `Local` works with both `Future` and Monix `Task/BIO`. 
+
+## Other effects
```
Comment (paualarco) [inline docs/overview.md:84]: This section might be in a better place at the end of the documentation? Just to make it shorter and let the user to go faster to hands on.
They could just skip it if they want to, but if I were the reader, I would expect to see some motivation at the start, and then at the end the comparison with other effects. 
Of course this is just personal opinion, if you prefer it here don't bother changing it ;)

Comment (Avasil) [reply docs/overview.md:84]: Do you mean to put it on a different page?

Comment (paualarco) [reply docs/overview.md:84]: Yup, I meant that, at the end of the doc indeed.

Comment (Avasil) [reply docs/overview.md:84]: I like the idea, thanks for the suggestion

File: docs/comparison.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+- *Fiber management.* Monix treats `Fiber` as a low level construct and recommends staying away from it unless you know what you're doing. 
+`ZIO` embraces the concept and adds fiber supervision, structured concurrency etc. Cool feature, but it's an extra thing to learn about, and I don't see many use cases for it.
+Usually it's possible to do everything with other concurrency combinators.
+- *Framework experience vs pluggable library.* ZIO is more opinionated and pushes you to use `ZIO`-everything, otherwise you might have an underwhelming experience.
+Monix wants to naturally interop with both FP and non-FP ecosystem but it might have less "batteries" included. 
+ZIO can be better at forcing you to follow their pattern of programming, while Monix can play nicer with your team's individual preferences and mixed (e.g. using both Monix and Future) codebases.
+
+### Summary
+
+ZIO has more contributors, more features which you may or may not use, is more opinionated and develops a new ecosystem of libraries which is tailored to ZIO needs.
```
Comment (Executioner1939) [inline docs/comparison.md:94]:
```suggestion
ZIO has more contributors, more features (which you may or may not use), is more opinionated, and develops a new ecosystem of libraries which are tailored to ZIO needs.
```

File: docs/overview.md [hunk 3]
```diff
@@ -3,4 +3,89 @@ id: overview
 title: Overview
 ---
 
-overview
\ No newline at end of file
+`BIO[E, A]` represents a specification for a possibly lazy or asynchronous computation. 
+When executed, it will produce a successful value `A`, an error `E`, never terminate or complete with a terminal (untyped) error.
+
+`BIO` handles concurrency, cancellation, resource safety, context propagation, error handling, and can suspend effects.
+All of it makes it a good tool to write a simple, high-level code to solve problems related to any of these features in a safe and performant manner.
```
Comment (Executioner1939) [inline docs/overview.md:10]:
```suggestion
All of this makes it simple to write good, high-level code that solve problems related to any of these features in a safe and performant manner.
```

File: docs/overview.md [hunk 4]
```diff
@@ -3,4 +3,89 @@ id: overview
...
+
+Well, it seems like built-in typed errors are in great demand, and the only other effect which has built-in typed errors is `ZIO`. 
+Not everyone wants to use `ZIO` and I feel like there are enough differences in `Monix` to make it a valuable alternative.
+The target audience of BIO are users of `IO`, `Task`, and `Future` who tend to use `EitherT` a lot and would like to have a smoother experience, with better type inference, no syntax imports, and without constant wrapping and unwrapping.
+
+`Monix BIO` builds upon `Monix Task` and enhances it with typed error capabilities.
+If you are already familiar with `Task` - learning `BIO` is straightforward because the only difference is in
+error handling - the rest of API is the same. In many cases, migration might be as simple as changing imports from `monix.eval.Task` to `monix.bio.Task`.
+
+To me, the big difference between Monix and other effect libraries is approach to impure code.
```
Comment (Executioner1939) [inline docs/overview.md:75]:
```suggestion
To me, the big difference between Monix and other effect libraries is it's approach to impure code.
```

File: docs/overview.md [hunk 5]
```diff
@@ -3,4 +3,89 @@ id: overview
...
+Both `IO` and `ZIO` will push you to write 100% purely functional codebase, except for isolated cases where low-level imperative code is needed for performance.
```
Comment (Executioner1939) [inline docs/overview.md:76]:
```suggestion
Both `IO` and `ZIO` will push you to write a 100% purely functional codebase, except for isolated cases where low-level imperative code is needed for performance.
```

File: docs/overview.md [hunk 6]
```diff
@@ -3,4 +3,89 @@ id: overview
...
+Monix favors purely functional programming too (the only impure method is `memoize`), but we also recognize and fully support users who would rather go for hybrid approach, or are allergic to pure FP.
```
Comment (Executioner1939) [inline docs/overview.md:77]:
```suggestion
Monix favors purely functional programming too (the only impure method is `memoize`), but we also recognize and fully support users who would rather go for a hybrid approach, or are allergic to pure FP.
```

File: docs/overview.md [hunk 7]
```diff
@@ -3,4 +3,89 @@ id: overview
...
+I feel like it plays to Scala's unique strengths. 
+Few examples of Monix providing extra support for `Future` users:
```
Comment (Executioner1939) [inline docs/overview.md:79]:
```suggestion
Here are a few examples of Monix providing extra support for users of `Future`:
```

File: docs/overview.md [hunk 8]
```diff
@@ -3,4 +3,89 @@ id: overview
...
+- `monix-execution` module provides many utilities to use with `Future` even if you're not interested in `Task` at all.
```
Comment (Executioner1939) [inline docs/overview.md:80]:
```suggestion
- The `monix-execution` module provides many utilities to use with `Future` even if you're not interested in `Task` at all.
```

File: docs/overview.md [hunk 9]
```diff
@@ -3,4 +3,89 @@ id: overview
...
+- Monix uses `Scheduler` which is also `ExecutionContext` and can be used with `Future` directly.
```
Comment (Executioner1939) [inline docs/overview.md:81]:
```suggestion
- Monix uses a `Scheduler` which is also an `ExecutionContext` and can be used with `Future` directly. 
```

File: docs/resourcesafety.md [hunk 10]
```diff
@@ -0,0 +1,219 @@
+---
+id: resource-safety
+title: Resource Safety
+---
+
+Many tasks use some kind of resource, such as a connection pool, a file handle, or a socket. 
+It's crucial to close them when we finish using them, otherwise we end up with resource leaks.
+
+`BIO` provides ways to do it safely and goes beyond capabilities of `try-catch-finally` block.
```
Comment (Executioner1939) [inline docs/resourcesafety.md:9]:
```suggestion
`BIO` provides ways to do it safely and goes beyond the capabilities of a `try-catch-finally` block.
```

File: docs/resourcesafety.md [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+
+## Running finalizer
+
+`guarantee` ensures that we run provided finalizer regardless of the exit condition, be it successful completion, failure, or cancellation.
```
Comment (Executioner1939) [inline docs/resourcesafety.md:13]:
```suggestion
`guarantee` ensures that we run the provided finalizer regardless of the exit condition, be it successful completion, failure, or cancellation.
```

File: docs/resourcesafety.md [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+  case ExitCase.Canceled => UIO(println("Task has been cancelled"))
+}
+  
+val fasterTask = UIO.evalAsync(10)
+  
+BIO.race(slowerTask, fasterTask).runSyncUnsafe()
+ //=> Task has been cancelled
+```
+
+`BIO.race` cancels slower task which executes the corresponding finalizer.
```
Comment (Executioner1939) [inline docs/resourcesafety.md:34]:
```suggestion
`BIO.race` cancels the slower task which executes the corresponding finalizer.
```

File: docs/resourcesafety.md [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+
+## Safe acquisition and release
+
+`bracket` is a more general operator for `try-with-resources` pattern, but it works for pure effect types like `BIO`,
```
Comment (Executioner1939) [inline docs/resourcesafety.md:38]:
```suggestion
`bracket` is a more general operator for the `try-with-resources` pattern, but it works for pure effect types like `BIO`,
```

File: docs/resourcesafety.md [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+  // Releasing the reader (the finally block)
+  val release: BufferedReader => UIO[Unit] = in => Task(in.close()).onErrorHandle(_ => ())
+
+  acquire.bracket(use)(release)
+}
+```
+
+If `Task` is successful, the result will be signaled in `use`.
+
+Note that `release` expects `UIO` - you can ignore any errors, or raise them as terminal errors but make sure to consider what should happen if finalizer fails.
```
Comment (Executioner1939) [inline docs/resourcesafety.md:60]:
```suggestion
Note that `release` expects `UIO` - you can ignore any errors, or raise them as terminal errors but make sure to consider what should happen if the finalizer fails.
```

File: docs/resourcesafety.md [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+Don't leak any resources!
+
+Similarly to `guarantee`, there is also `bracketCase` variant.
```
Comment (Executioner1939) [inline docs/resourcesafety.md:63]:
```suggestion
Similarly to `guarantee`, there is also a `bracketCase` variant.
```

File: docs/resourcesafety.md [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
+  acquire.bracketCase(use)(release)
+}
+```
+
+## cats.effect.Resource
+
+[Resource](https://typelevel.org/cats-effect/datatypes/resource.html) is compatible with `BIO`.
+It allows us to define a task which has already specified `acquire` and `release` logic and clearly informs the users that they are not responsible for closing the resource.
+
+Let's take a look at the example from Cats-Effect docs, but written in `BIO`:
```
Comment (Executioner1939) [inline docs/resourcesafety.md:91]:
```suggestion
Let's take a look at the example from the Cats-Effect docs, but written in `BIO`:
```

File: docs/resourcesafety.md [hunk 17]
```diff
@@ -0,0 +1,219 @@
...
+
+Resource.make(acquire)(release).evalMap(addDogs).use(report).runSyncUnsafe()
+//=> Acquire cats...
+//=> ...more animals...
+//=> ...produce weather report...
+//=> ...release everything
+// Returns "It's raining cats and dogs"
+```
+
+The main drawback is that `Resource.use` requires error type to be `Throwable` but we can work with typed errors for `acquire` and `release`.
```
Comment (Executioner1939) [inline docs/resourcesafety.md:113]:
```suggestion
The main drawback is that `Resource.use` requires the error type to be `Throwable` but we can work with typed errors for `acquire` and `release`.
```

File: docs/resourcesafety.md [hunk 18]
```diff
@@ -0,0 +1,219 @@
...
+val use: String => BIO[String, String] = s => BIO.now(s)
+val release: String => UIO[Unit] = _ => UIO (throw DummyException("unexpected error"))
+
+val task: BIO[String, String] = acquire.bracket(use)(release)
+val result: Either[String, String] = task.attempt.runSyncUnsafe()
+
+//=> Exception in thread "main" monix.execution.exceptions.DummyException: unexpected error
+```
+
+If both `use` and `release` fails, the errors are merged via `Platform.composeErrors` and signalled as a terminal failure:
```
Comment (Executioner1939) [inline docs/resourcesafety.md:136]:
```suggestion
If both `use` and `release` fail, the errors are merged via `Platform.composeErrors` and signalled as a terminal failure:
```

File: docs/resourcesafety.md [hunk 19]
```diff
@@ -0,0 +1,219 @@
...
+  .guarantee(UIO(println("finalizer C")))
+  .runSyncUnsafe()
+
+//=> action
+//=> finalizer A
+//=> finalizer B
+//=> finalizer C
+```
+
+All finalizers will be executed even if any of them fails:
```
Comment (Executioner1939) [inline docs/resourcesafety.md:177]:
```suggestion
All finalizers will be executed even if any of them fail:
```

**Comment by Avasil** [comment]
Thanks for the feedback @Executioner1939 and @paualarco 

I have rewritten the `Motivation` section - I would love to hear your thoughts.
I'm trying to convey how I see Monix in the ecosystem, but I'm a bit afraid I give an impression like Monix BIO is not purely functional. I don't want to scare fellow FP enthusiasts. :D

It would be also good to know if you have a completely view of Monix.

**Comment by paualarco** [comment]
@Avasil it's a good introduction actually! Looking forward to see more documentation to learn more :)
Although you are giving personal opinion, I would suggest avoiding expressions like `To me, the big difference ` generally speaking in first person :P 
You would also like to fix the format of this part `to make it a valuable alternative. For instance, if you are a happy user of Typelevel libraries (http4s, fs2, doobie etc.) you might find that`.

**Comment by paualarco** [comment]
> It would be also good to know if you have a completely view of Monix.
Did you mean that if overview section of documentation gives you a good view on why /when to use Bio vs Monix?

**Comment by Avasil** [comment]
> Although you are giving personal opinion, I would suggest avoiding expressions like To me, the big difference generally speaking in first person :P

Yeah, I agree, it's not a blog post. I'll make another pass, maybe I can rephrase it


> > It would be also good to know if you have a completely view of Monix.
> 
> Did you mean that if overview section of documentation gives you a good view on why /when to use Bio vs Monix?

No, in `Monix Niche` section I am writing about Monix in general. 
It would be better to put it in a shared place in docs for all of Monix, but it's not organized like that atm, and I wanted to mention it. I hope it's not too confusing in the current state.

Currently I want to write perhaps basic, but complete documentation for `monix-bio` that can be understood in isolation (without knowing about any effect types). 

In the next step, I would like to share `BIO` and `Task` docs because everything apart from error handling could be almost the same. I wonder if it's hard to do something like Spark with different APIs: https://spark.apache.org/docs/latest/quick-start.html but I can always copy paste. :D Current documentation of `monix.eval.Task` is a few years old, and I want to give it a revamp along the way. 

Then if I'm not completely burned out, I'll attack `Observable`. Although I would like to create a sample app at some point so it will take me a while :D

**Comment by paualarco** [comment]
> No, in `Monix Niche` section I am writing about Monix in general.

I think it is good, but agree it is not that related with monix-bio, but with monix in general.

> 
> In the next step, I would like to share `BIO` and `Task` docs because everything apart from error handling could be almost the same. 

IMO `monix-bio` looks very close to be another sub-module of monix and not as a different project. But at the same time it is not that mature. 

So, I just agree that would make lot of sense merging task and bio docs, and that would be a good opportunity to start approaching the monix future plans of moving documentation to a better organised format (docusaurus).  

> 
>I wonder if it's hard to do something like Spark with different APIs: https://spark.apache.org/docs/latest/quick-start.html but 
>I can always copy paste. :D 

That would be awesome but no idea how to do so atm ;D

**Pull request merged**